### PR TITLE
[MER-140] Add generated routes

### PR DESCRIPTION
**Pull request by kizza**
- Adds generated routes to static site generation
- Adds generated routes to express server
- Passes in env variables to client bundle

### CODE REVIEW

File: source/server.js [hunk 1]
```diff
@@ -5,12 +5,26 @@ import { match, RouterContext } from 'react-router'
...
 import Document from './layouts/Document'
 import routes from './Routes'
+import { fetchCharities } from './store/actions/charities'
 import { configureStore } from './store'
 import { trigger } from 'redial'
 
-export default (route, callback) => {
-  const store = configureStore()
+const prefetchedStore = () => (
+  new Promise((resolve, reject) => {
```
Comment (bradparker) [inline source/server.js]: Don't _need_ to wrap this one up in a promise if you don't want.

File: build/server.js [hunk 2]
```diff
@@ -12,16 +12,20 @@ const server = express()
...
+          if (err) { return res.send(`<h1 style="color:red">${err.message}</h1>\n<p>${err.stack}</p>`) }
+          res.send(html)
+        })
+      })
     })
   })
-})
 
 module.exports = (callback) => (
   http.createServer(server).listen(8080, callback)
```
Comment (bradparker) [inline build/server.js]: We might want to do the route fetching within this exported function...
We have a nifty callback being passed along which we can use to be sure that the server has had the chance to have its routes defined before we try to start it.

**Comment by bradparker** [comment]
Looks pretty damn good.
Awesome

File: source/server.js [hunk 3]
```diff
@@ -5,12 +5,23 @@ import { match, RouterContext } from 'react-router'
...
 import routes from './Routes'
+import { fetchCharities } from './store/actions/charities'
 import { configureStore } from './store'
 import { trigger } from 'redial'
 
-export default (route, callback) => {
+const prefetchedStore = () => {
   const store = configureStore()
+  return fetchCharities(store.dispatch)()
+    .then(() => (Promise.resolve(store)))
```
Comment (bradparker) [inline source/server.js]: Blerg, promises are weird. But you can just return the store in the `then`. The result it the same.

```
> const prom = new Promise((resolve, reject) => { resolve('foo') }).then(() => 'bar')
undefined
> prom.then((r) => console.log(r))
Promise { <pending> }
> bar
> prom.then((r) => console.log(r))
> const prom2 = new Promise((res, rej) => { res('foo') }).then(() => Promise.resolve('bar'))
undefined
> prom2.then((r) => console.log(r))
Promise { <pending> }
> bar
```

Comment (kizza) [inline source/server.js]: Yeah, I did think that - swear I had it return an error for not having "then()" but yes, you're right! Changing...

File: source/Routes/FeatureCharity/index.js [hunk 4]
```diff
@@ -37,12 +37,14 @@ const hooks = {
   fetch: ({
     dispatch,
     params,
-    campaign
+    campaign,
+    charities
```
Comment (kizza) [inline source/Routes/FeatureCharity/index.js]: @brad correct me if i'm wrong but I believe we need to pass this in from state (featured charity would always fetch charities)

Comment (bradparker) [inline source/Routes/FeatureCharity/index.js]: Hrm, now we weren't fetching them if we already had them. I think it'd be nice to preserve that behaviour.

That way we can rely on webhook-triggered rebuilds keeping the content up to date

File: source/server.js (hunk 3 above)
Comment (bradparker) [inline source/server.js]: One more item to add to the "promises are weird" list:
If you're function returns undefined it _wont_ return a new promise.... I think in that case the `.then` should just add another subscriber... anyway, promises are weird.

Comment (kizza) [inline source/server.js]: Huh!  Perhaps that's what I ran into :+1:

**Comment by bradparker** [comment]
Ok, last thing! I'll get the tests from the base branch passing and then we should be good

**Pull request merged**